feat(aws): support multiple nodes

FrankYang0529 opened the pull request:
ref: https://github.com/fermyon/installer/issues/63

FrankYang0529 commented:
Hi @vdice, sorry for pushing a draft PR first. I encountered an issue and would like to get your feedback first. Thank you!

Currently, this branch can deploy 3 instances, hasicorp stacks, and fermyon platform. However, when I tried to run `spin deploy` to deploy [http-rust](https://github.com/fermyon/spin/tree/main/examples/http-rust). I always get an error from bindle. I'm not familiar with bindle part. Could you help me take a look for it. Thank you!

```
> spin --version
spin 0.4.0 (784094e 2022-07-13)

> spin deploy
Error: Failed to push bindle spin-hello-world/1.0.0+q8e6334c to server https://bindle.44.210.10.26.sslip.io/v1

Caused by:
    0: Failed to push bindle from '/var/folders/jp/2ylhtk2x0m1fgp3s7k0yb3kr0000gn/T/.tmpoJwMBS' to server at 'https://bindle.44.210.10.26.sslip.io/v1'
    1: Error contacting server: resource could not be loaded
```

For design, originally, I would like to use `aws_lb`. However, ALB DNS record can't support wildcard subdomain, so we need users to provide their domain name and add a A record to ALB. After realizing it, I choose current simple way for it - using `sslip.io` on `eip`. I will add ALB support after I resolve `spin deploy` issue.

Also, I wonder how to sync host volume between nodes? If we can't do it, will Postgres be redeployed to another node? If yes, we may lose data in the last node.

For vault, I am not sure where we use it. I don't deploy it in current version. Will it break something?

For consul/nomad, do we want to split client nodes to other instances?

vdice commented:
@FrankYang0529 no worries; great idea to put up as draft first anyways on a big effort like this.  I hope to deploy from this branch  and take a look at the bindle issue soon.

Meanwhile:

>  I will add ALB support after I resolve spin deploy issue.

👍 

> Also, I wonder how to sync host volume between nodes? If we can't do it, will Postgres be redeployed to another node? If yes, we may lose data in the last node.

Right, we would lose data with the current postgres job config.  Good question, we'd need to research techniques (use AWS EFS? Are there similar options for other cloud installers (GCP, DO, etc.)?  Is there a cloud platform-agnostic approach?)  Maybe relegate to a follow-up due to its complexity and reach across installer scenarios?

> For vault, I am not sure where we use it. I don't deploy it in current version. Will it break something?

It is fine to leave Vault out of the mix for now -- you're correct, it isn't used corrently.

> For consul/nomad, do we want to split client nodes to other instances?

Perhaps eventually, but would be fine to tackle in a different issue.  For now, I like the simplicity of each node looking like a clone of the rest -- though we'll see if that bites us somehow as we test :)

FrankYang0529 commented:
> Is there a cloud platform-agnostic approach?

I took a look for [ceph](https://docs.ceph.com/en/latest/rbd/rbd-nomad/) today, but it looks like we need a ceph cluster first. I don't want to introduce too many components in this PR, so probably let's use [EBS](https://learn.hashicorp.com/tutorials/nomad/stateful-workloads-csi-volumes?in=nomad/stateful-workloads#deploy-an-aws-ebs-volume) first? From [nomad document](https://learn.hashicorp.com/tutorials/nomad/stateful-workloads-csi-volumes?in=nomad/stateful-workloads), we can also use a similar technique in GCP and Digital Ocean. (GCP persistent disks and Digital Ocean droplet storage volumes)

vdice commented:
> I took a look for [ceph](https://docs.ceph.com/en/latest/rbd/rbd-nomad/) today, but it looks like we need a ceph cluster first. I don't want to introduce too many components in this PR

Agreed, let's try using the EBS/container storage interface approach first.  Good find.

> I always get an error from bindle. I'm not familiar with bindle part. Could you help me take a look for it. Thank you!

Ok, I had a chance to test this.  The bindle issue is due to bindle not being able to write to the host volume.  There are two changes needed; I'll add comments in-line.

Once I applied the changes around the bindle volume, I was able to attempt a `spin deploy`.  I saw that the bindle transaction was successful (bindle shows up in volume) but it looks like our next hurdle is related to hippo's postgresql driver and the postgres container; I see the following: https://gist.github.com/vdice/cc081811e1649c3cb3a5abf55bd71b5f

FrankYang0529 commented on aws/terraform/vm_assets/job/bindle.nomad:
```diff
@@ -0,0 +1,77 @@
+    volume "bindle" {
+      type      = "host"
+      source    = "bindle"
+      read_only = true
```
Thanks for catching this!! I will update it.

FrankYang0529 replied:
Fixed it. Thank you.

FrankYang0529 commented:
> Once I applied the changes around the bindle volume, I was able to attempt a `spin deploy`. I saw that the bindle transaction was successful (bindle shows up in volume) but it looks like our next hurdle is related to hippo's postgresql driver and the postgres container; I see the following: https://gist.github.com/vdice/cc081811e1649c3cb3a5abf55bd71b5f

I tested different Hippo versions. It looks like there is an issue to work with PostgreSQL. I create an issue to track it.

https://github.com/deislabs/hippo/issues/1167

FrankYang0529 commented on aws/terraform/scripts/install_nomad.sh.tpl:
```diff
@@ -0,0 +1,58 @@
+
+# create folders
+sudo mkdir -p /opt/bindle/data
+sudo chown nomad:nomad /opt/bindle/data
```
Updated it. Thank you.

FrankYang0529 commented on aws/terraform/scripts/install_nomad.sh.tpl:
```diff
@@ -0,0 +1,58 @@
+sudo mkdir -p /opt/bindle/data
+sudo chown nomad:nomad /opt/bindle/data
+sudo mkdir -p /opt/postgres/data
+sudo chown nomad:nomad /opt/postgres/data
```
Removed postgres folder and used CSI.

FrankYang0529 commented:
Hi @vdice, I fixed the error config for bindle and updated postgres host volume to aws-ebs CSI. I think we can also dockerize `bindle`, so we don't need to use host volume for it. I would like to leave `aws_elb` to the next PR, so the remaining items for this PR are:

* [ ] Dockerize bindle
* [ ] Update Hippo/Spin version
* [ ] Update README

vdice commented:
@FrankYang0529 the ebs setup looks slick!  I was encountering errors when testing -- it seems like it would schedule all of the jobs up to and including postgres but hippo never ran... however, I thought maybe some things are still in progress so didn't dig deep.

👍 on list from https://github.com/fermyon/installer/pull/100#issuecomment-1242927570

FrankYang0529 commented on aws/terraform/variables.tf:
```diff
@@ -1,19 +1,43 @@
+  default     = "t2.medium"
+}
+
+variable "availability_zone" {
```
Add `data "aws_region"` and remove the variable. I think we can use this currently and add the variable back in the follow-up PR. Thank you.

FrankYang0529 commented:
Hi @vdice, I did some updates, but haven't fully tested it. I just wanted to keep some steps that I used in manual deployment. I will do more tests tomorrow. Thank you.

* Use docker driver for bindle.
* Generate bindle key in the script.
* Only run `run_servers.sh` in the first node.

FrankYang0529 commented:
Hi @vdice, I tested the provisioning part. Currently, it can deploy all components successfully. However, we need a new Hippo release to avoid the PostgreSQL DateTime issue https://github.com/deislabs/hippo/issues/1167. Also, we need a CI update to publish an official bindle image. I used `frankyang/bindle:dev` for testing. I will create a PR in bindle repo for it.

bacongobbler commented:
Thanks @FrankYang0529. I'll see about cutting a new Hippo release some time this week.

bacongobbler commented:
https://github.com/fermyon/installer/pull/109

FrankYang0529 commented:
Besides README, this PR is almost ready. Currently, it waits for https://github.com/fermyon/spin/pull/786 and https://github.com/deislabs/bindle/pull/353, so I would like to add some test steps for anyone who wants to try this.

1. Apply resources
```bash
$ cd aws/terraform
$ terraform apply
```
2. Export ssh key
```bash
$ terraform output -raw ec2_ssh_private_key > /tmp/ec2_ssh_private_key.pem
$ chmod 600 /tmp/ec2_ssh_private_key.pem
```
3. Login to ec2 and wait for cloud-init to finish.
```bash
$ ssh -i /tmp/ec2_ssh_private_key.pem ubuntu@<elastic_ip>
$ sudo journalctl -u cloud-final.service -f
```
4. After cloud-init is finished, you can check nomad jobs on `https://<elastic_ip>:4646`.
5. Copy `secret_keys.toml` and `keyring.toml` from ec2 to your laptop.
```bash
# on ec2
$ cd ~
$ cat secret_keys.toml
$ cat keyring.toml
```
6. Clone and build https://github.com/fermyon/spin/pull/786.
```
# in spin folder
$ cargo build
$ ./target/debug/spin templates install --git https://github.com/fermyon/spin
```
7. Run the following commands in the same terminal.
```bash
# in installer/aws/terraform folder
$ $(terraform output -raw environment)
# in spin folder
$ ./target/debug/spin new http-rust myapp
$ export BINDLE_KEYRING_FILE=<path to keyring.toml>
$ export BINDLE_SECRET_FILE=<path to secret_keys.toml>
$ export BINDLE_LABEL="user<me@example.com>"
$ cd myapp
$ ../target/debug/spin build
$ ../target/debug/spin deploy
```
8. Open `https://spin-deploy.myapp.hippo.<elastic_ip>.sslip.io` and you will see `Hello, Fermyon`.

If you do some change and want to push the app again, you can do following commands with same terminal for step 7. We need environment variables from `$(terraform output -raw environment)` and `BINDLE_KEYRING_FILE`, `BINDLE_SECRET_FILE`, and `BINDLE_LABEL`.
9. Install [bindle `v0.9.0-rc1`](https://github.com/deislabs/bindle/releases/tag/v0.9.0-rc.1) on your local.
10. Fetch server keys. In stdout, you will see where it write the output to.
```bash
$ bindle keys fetch
```
11. Add the key to your `BINDLE_KEYRING_FILE`.
12. In myapp folder, push the app again.
```bash
$ ../target/debug/spin build
$ ../target/debug/spin deploy --deploy-existing-bindle
```

vdice commented:
Awesome work, @FrankYang0529!

I have yet to test the flow laid out in https://github.com/fermyon/installer/pull/100#issuecomment-1304811993 but I do want to mention the following item to get your thoughts:

Can we extract the bindle version update (from v0.8.0 to v0.9.x) into a separate follow-up?  I ask for a few reasons:

1. When we're ready to do so, I'd like the bindle version to be updated in all scenarios (local and cloud providers) at once (eg same PR), if possible.
2. It would be great to scope this PR down to the changes only needed to support multi-node clusters in AWS
3. It would be preferable not to be tightly-coupled to the correspoding Spin PR that bump its bindle client version (https://github.com/fermyon/spin/pull/786)

For 2 (the purposes of this PR), do we really only need a publicly-accessible Docker image of bindle, regardless of version?  If we had such an image for bindle v0.8.0, can we still achieve running the platform in AWS in a multi-node context -- and use the currently-available v0.6.0 spin to deploy?

What do you think?  I know you've spent a lot of time on this and so I may be missing some details...

FrankYang0529 commented:
Hi @vdice, yeah, I think it's easier to use Bindle v0.8.x. Let's do a follow-up in the next PR. I also update README.

> If we had such an image for bindle v0.8.0, can we still achieve running the platform in AWS in a multi-node context -- and use the currently-available v0.6.0 spin to deploy?

Yes, we can. Since we don't use bindle v0.9.x, all test steps are the same as usual. Currently, I temporally used bindle image `frankyang/bindle:v0.8.3`. If we merge https://github.com/fermyon/bindle/pull/2 and release a new `fermyon/bindle` image, I will change it to `ghcr.io/fermyon/bindle:v0.8.3`. Thanks

FrankYang0529 commented:
Hi @vdice, I updated the bindle image to `ghcr.io/fermyon/bindle:v0.8.2` and it works fine. This PR is ready for review. Thanks.

vdice commented:
@FrankYang0529 thanks so much; sorry for the delay here.  Should have cycles to devote to this tomorrow -- excited to test it out!

vdice commented on aws/terraform/vm_assets/run_servers.sh:
```diff
@@ -0,0 +1,137 @@
+sudo mount -t ext4 -o defaults /dev/xvdb /bindle
+sudo chown ubuntu:ubuntu /bindle
+
+# echo "generate bindle key"
```
We can remove these...

vdice commented on aws/terraform/vm_assets/run_servers.sh:
```diff
@@ -0,0 +1,137 @@
+echo "----------"
+echo "Consul:  http://${IP_ADDRESS}:8500"
+echo "Nomad:   http://${IP_ADDRESS}:4646"
+# echo "Vault:   http://localhost:8200"
```
Let's remove these commented-out Vault lines...

FrankYang0529 replied:
Removed it. Thanks.

vdice commented on aws/terraform/vm_assets/run_servers.sh:
```diff
@@ -0,0 +1,137 @@
+
+trap cleanup EXIT
+
+nomad_run() {
```
Curious to hear the reason this was needed -- was it around specific jobs?  What was the error(s) seen?

FrankYang0529 replied:
I remembered I got `Error fetching deployment` sometimes and we don't have `retry` feature in nomad now.

ref: https://github.com/hashicorp/nomad/issues/12062

vdice replied:
Ah, I see.  I think I've encountered this as well.  Thank you for the link

FrankYang0529 commented:
> However, adding logic to support both modes may be complex/involved and I don't want to derail the progress here. What do you think? Willing to be convinced that we just drop the single-node formation, pending user feedback -- we can always bring back from git history...

How about let's split folders for single-node and multiple-node deployment? This may have more work for maintenance, but we can share single-node scripts with GCP/Azure/DO, so I think it's acceptable.

FrankYang0529 commented on aws/README.md:
```diff
@@ -5,10 +5,11 @@ This guide illustrates how to install Fermyon on AWS using Terraform.
+on multiple EC2 instances. The data is stored on AWS EBS.
 
+Since the infrastructure uses a single node with elastic ip to receive requests,
+it might have single point failure.
```
Fixed it. Thank you.

FrankYang0529 commented on aws/README.md:
```diff
@@ -35,20 +36,30 @@ is enabled, apps will be provided with https URLs and TLS certs courtesy LE.
-    - Name: `${var.instance_name}`
-  - 1 Elastic IP address (associated with instance)
-    - Name: `${var.instance_name}-eip`
+  - 3 EC2 instance (default size: `t2.medium`)
```
Updated sentence because we have both single-node and multiple-nodes folder.

FrankYang0529 commented on aws/terraform/main.tf:
```diff
@@ -3,10 +3,12 @@
 
 locals {
-  dependencies = yamldecode(file("../../share/terraform/dependencies.yaml"))
+  dependencies = yamldecode(file("./dependencies.yaml"))
```
Using sharing dependencies.yaml.

FrankYang0529 commented on aws/terraform/main.tf:
```diff
@@ -135,31 +279,152 @@ resource "aws_instance" "ec2" {
+
+  # Hippo port
+  ingress {
+    description = "Nomad dynamic ports"
```
Updated it. Thank you.

FrankYang0529 commented on aws/terraform/main.tf:
```diff
@@ -135,31 +279,152 @@ resource "aws_instance" "ec2" {
+    from_port   = 5000
+    to_port     = 5000
+    # TODO: change cidr
+    cidr_blocks = ["0.0.0.0/0"]
```
Yes, updated it. Thank you.

FrankYang0529 commented on aws/terraform/scripts/install_consul.sh.tpl:
```diff
@@ -0,0 +1,39 @@
+echo "Installing Consul"
+sudo apt-get install consul
```
Changed to use the version in shared dependencies.yaml.

FrankYang0529 commented on aws/terraform/scripts/install_nomad.sh.tpl:
```diff
@@ -0,0 +1,83 @@
+echo "Installing Nomad"
+sudo apt-get install nomad
```
Changed to use the version in shared dependencies.yaml.

FrankYang0529 commented on aws/README.md:
```diff
@@ -35,20 +36,30 @@ is enabled, apps will be provided with https URLs and TLS certs courtesy LE.
-    - Name: `${var.instance_name}-security-group`
-    - Inbound connections allowed for ports 22, 80 and 443
+    - Name: `${var.resource_name_prefix}-server-sg`
+    - Inbound connections allowed for ports 22, 80, 443, 8081 (traefik), 4646 (nomad), and 8500 (consul)
```
Fixed it. Thank you.

FrankYang0529 commented on aws/terraform/variables.tf:
```diff
@@ -1,13 +1,37 @@
+}
+
+variable "bindle_disk_size" {
+  description = "Bindle disk size (Default: 20)"
```
Fixed it. Thank you.

FrankYang0529 commented on aws/terraform/variables.tf:
```diff
@@ -1,13 +1,37 @@
+}
+
+variable "postgres_disk_size" {
+  description = "Postgres disk size (Default: 20)"
```
Fixed it. Thank you.

FrankYang0529 commented on aws/terraform/variables.tf:
```diff
@@ -1,13 +1,37 @@
+  default     = 20
+}
+
+variable "bindle_label" {
```
Commented it. Thank you.

FrankYang0529 commented on aws/terraform/vm_assets/job/hippo.nomad:
```diff
@@ -0,0 +1,142 @@
+# TODO: deploy postgresql
```
Removed it. 😆

FrankYang0529 commented on aws/terraform/vm_assets/job/hippo.nomad:
```diff
@@ -0,0 +1,142 @@
+        Hippo__Administrators__0__Username = var.registration_mode == "AdministratorOnly" ? var.admin_username : ""
+        Hippo__Administrators__0__Password = var.registration_mode == "AdministratorOnly" ? var.admin_password : ""
+
+        # Database Driver: inmemory, sqlite, postgresql
```
Removed it. Thank you.

FrankYang0529 commented on aws/terraform/vm_assets/job/hippo.nomad:
```diff
@@ -0,0 +1,142 @@
+      template {
+        data = <<EOF
+{{ range service "postgres" }}
+ConnectionStrings__Database = "Host={{ .Address }}:{{ .Port }};Username=postgres;Password=postgres;Database=hippo"
```
Added a TODO for it. We can fix it in follow-up.

vdice commented on aws/terraform/main.tf:
```diff
@@ -135,31 +279,152 @@ resource "aws_instance" "ec2" {
+    protocol    = "TCP"
+    from_port   = 20000
+    to_port     = 32000
+    # TODO: change cidr
```
Traefik is the proxy for all requests to Spin apps (and hippo and bindle), so only the Traefik ports (80/443) should need to be publicly accessible (which they are).  I switched this to `self = true` and am able to reach my deployed spin app.  Or were you encountering issues?

vdice reviewed (commented):
Just tested the `multiple-nodes` scenario on AWS and the services came up as expected and an example app deploy worked! 🎉 

I think we're at the stage of making sure the docs are in order (and accurate to the scenario, now that we have multi- and single- modes).  Then we should be good to merge this in and start using.

The multiple-nodes configuration is a more robust and flexible foundation compared to the single-node configuration, so I'm excited to iterate here!

vdice commented on aws/terraform/multiple-nodes/main.tf:
```diff
@@ -0,0 +1,510 @@
+  security_group_id = aws_security_group.server_sg.id
+}
+
+resource "aws_security_group_rule" "allow_nomad_api_inbound" {
```
I'm seeing weird behavior when I allow inbound nomad/consul api to my cluster like so:
```
terraform apply -var allow_inbound_http_nomad=true -var allow_inbound_http_consul=true
```
When I do so, it seems to _remove_ other rules in the security group, for ports 22, 80, 443.

I think there may be some issue with merging these `aws_security_group_rule` resources with the main `aws_security_group` object... can you also try and see what you encounter?

FrankYang0529 replied:
I didn't encounter this issue. Ports 22, 80, 4646, 5000, 8081, and 8500 are on my inbound list.

<img width="961" alt="Screenshot 2022-11-23 at 11 22 34 PM" src="https://user-images.githubusercontent.com/4927361/203585664-7dc7cd70-fe35-42f2-9925-2179823c892d.png">

vdice replied:
Ok, I think I've pinned it down.  The odd behavior occurs when running `terraform apply` on an existing installation.  It looks like indeterministic behavior results when we mix an `aws_security_group` resource with in-line `ingress` rules in tandem with `aws_security_group_rule` resources; see https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/security_group_rule

<img width="659" alt="Screenshot 2022-11-23 at 10 46 11 AM" src="https://user-images.githubusercontent.com/9322017/203624865-7045c99f-7f6f-45ee-98bc-f8f1d3c7040c.png">

FrankYang0529 replied:
Thanks for catching this! Updated it. I used `terraform apply` and then `terraform apply -var allow_inbound_http_nomad=true -var allow_inbound_http_consul=true`. It can add security group rule successfully.

FrankYang0529 commented on aws/README.md:
```diff
@@ -35,20 +36,36 @@ is enabled, apps will be provided with https URLs and TLS certs courtesy LE.
-    - Name: `${var.instance_name}`
-  - 1 Elastic IP address (associated with instance)
-    - Name: `${var.instance_name}-eip`
+  - EC2 instance
```
Used individual README.md files for two folders.

vdice replied:
Thank you.  On last item here: Can we update the top-level README to point to the two scenarios?  https://github.com/fermyon/installer#aws

FrankYang0529 replied:
Fixed it. Thank you.

vdice commented on aws/terraform/single-node/main.tf:
```diff
@@ -3,7 +3,7 @@
 
 locals {
-  dependencies = yamldecode(file("../../share/terraform/dependencies.yaml"))
+  dependencies = yamldecode(file("../../../share/terraform/dependencies.yaml"))
```
There are a few other spots in this file where this path also needs updating, such as [here](https://github.com/fermyon/installer/blob/4da0371441b5a570b4b5c7c59045006f3bd2d0ad/aws/terraform/single-node/main.tf#L78) and [here](https://github.com/fermyon/installer/blob/4da0371441b5a570b4b5c7c59045006f3bd2d0ad/aws/terraform/single-node/main.tf#L78).

FrankYang0529 replied:
Updated it. Thank you!

vdice commented on README.md:
```diff
@@ -34,7 +34,7 @@ apps.
 ---
 
-The [AWS Quick-start](./aws/README.md) is a great option to go if you'd like
+The [AWS single node](./aws/single-node/README.md) and [AWS multiple nodes](./aws/multiple-nodes/README.md) are great options to go if you'd like
```
How about something like, 

_The [AWS single node](./aws/single-node/README.md) and [AWS multiple node](./aws/multiple-nodes/README.md) scenarios are great options to go if you'd like to run Fermyon in a separate environment.  The former is quicker to launch and uses less cloud resources, while the latter is more robust and enables scaling the cluster size to suit your needs._

FrankYang0529 replied:
Updated it! Thank you. 🚀

vdice reviewed (commented):
Tested both single- and multi- scenarios and they look great.  Last minor docs suggestion and then I believe we're ready to get this in! 🎉

vdice reviewed (approved):
Thank you for the incredible amount of work here @FrankYang0529!  Super excited to now have a multi-node scenario for users to try.  🚀
